src/main.rs

```diff
@@ -44,10 +46,13 @@ async fn main() -> anyhow::Result<()> {
 
     match &command {
         Some(Commands::Manifests) => {
-            println!(
-                "{}",
-                serde_yaml::to_string(&kubit::resources::AppInstance::crd()).unwrap(),
-            );
+            let f = OpenOptions::new()
+                .write(true)
+                .open(kubit::resources::APPINSTANCE_CRD_PATH)
+                .expect("Could not open AppInstances CRD file");
+
+            serde_yaml::to_writer(&f, &kubit::resources::AppInstance::crd()).unwrap();
```

Comment:
yeah, a (possibly required) flag with the destination directory is the way to go.

As, for iterating, let's start manually.

---

`OpenOptions` is the general tool to handle file opening, but for simple operations is more common to just use `File::create`.

The OpenOptions builder is useful but its also tricky to use correctly. In fact you've used it incorrectly: you'll overwrite the file but you won't truncate the file and thus there will be trailing garbage if the new output is shorter than the existing contents of the file. See doc:

> If the file already exists, any write calls on it will overwrite its contents, without truncating it.

---

Minor details: rust naming convention unlike Go dislikes single letter names in most cases.

Reply:
> In fact you've used it incorrectly: you'll overwrite the file but you won't truncate the file and thus there will be trailing garbage

I just ran into this and was wondering why, this answers it :laughing: I was getting a trailing `s: {}` on files after I was running it a couple of times.

Reply:
other building blocks with which you can build a more generic code:

1. `format!(...)` is like `print!()` but returns a string. You can use it to construct filenames from parts
2. the `crd()` method returns a CustomResourceDefinition object, which is just a k8s resource. It has a `spec` which has a `group` and has a `names` which has a `plural`... . With that you can easily construct the `"kubecfg.dev_appinstances.yaml"` string.
3. you can start by iterating over a literal list, possibly of length one like `for crd in vec![&kubit::resources::AppInstance::crd()] {`
4. to construct paths from dir + filename, use `PathBuf`. The `PathBuf::from("a/dir").join("a_file")` construct is useful.

Reply:
another tip:

rust enum "constructors" (the variants inside then enum) come in three kinds:

* unit-like: `Foo`
* tuple-like: `Bar(String, i32)`
* struct-like `Baz{ a: String, b: i32 }`

The clap-derive crate offers a very nice way to declare flags and the structures that capture the flag values at the same time.

We currently have a single unit-like in the Commands enum constructor:

```rust
   #[derive(Clone, Subcommand)]
    enum Commands {
        /// Generates k8s manifests
        Manifests,
    }
```

This basically tells clap that there is a `manifests` subcommand that takes no args and no flags.
 
If you turn this into a tuple-like enum you'll tell clap that you expect an arg. But we want a flag.
To tell it it's a flag you need to annotate the field with `#[clap(long, default_value = ....)]` but you can do that only in the struct-like kind